### PR TITLE
Camera control from the game

### PR DESCRIPTION
**Pull request by Bryukh**
# Issue

We need to control the camera (bounds, target, zoom) from the game codebase.

# Fix

Added world's method to trigger client code events from the world lib.
Unbound the camera from the hero thang on game-dev levels (we don't hero there).

### CODE REVIEW

**Comment by AndrewJakubowicz** [comment]
If I understand correctly, this means you can publish the "camera:zoom-to" command from the game database?
Looks great.
This method allows the game codebase to publish any event. Could this be abused? Why not attach a public camera method to the world with a zoom to method?

**Comment by Bryukh** [comment]
> If I understand correctly, this means you can publish the "camera:zoom-to" command from the game database?

Yes.

> This method allows the game codebase to publish any event. Could this be abused?

The same way as you can do it from browser dev console I think. I can exec for example "Backbone.Mediator.publish("camera:set-camera", {bounds: [{x: 0, y: 0}, {x: 240, y: 160}]})" from the dev console and get the same effect as `publishGlobalEvent`. So if it can be abused then we have more serious problems :-)

> Why not attach a public camera method to the world with a zoomTo method?

This what I tried to do first, however `world` that we have in game codebase is not the same `world` that we have in `Surface`, and game `world` don't have direct access to surface and camera.

Plus this way we get more control then certain methods.

> And does freeing camera from game dev set the camera to defaults?

Without "freeing" camera its `zoomTo` HeroPlaceholder on each frame, but in game dev levels HeroPlaceholder is gameReferee so when I'm trying to zoom to a thang on next frame the camera move to a corner (if it's zoom in)

I need to think about "abuse" hypotesis and check one idea, maybe it's not the best way. Thanks!

**Comment by Bryukh** [comment]
@AndrewJakubowicz 
I've changed the publish event method and allowed only a camera event. I see how publishing any event can be abused. Thank you for that your point, you are right.

File: app/lib/world/world.coffee [hunk 1]
```diff
@@ -378,6 +378,12 @@ module.exports = class World
     return unless @goalManager
     @goalManager.submitWorldGenerationEvent(channel, event, @frames.length)
 
+  publishCameraEvent: (eventName, event) ->
+    return if not Backbone?.Mediator # headless mode don't have this
+    event ?= {}
+    channel = 'camera:' + eventName
```
Comment (AndrewJakubowicz) [inline app/lib/world/world.coffee]: Should this be passed in as the first argument beneath?
Or maybe `eventName = 'camera:' + eventName` could also work.

**Comment by AndrewJakubowicz** [comment]
This is great! Thanks for answering my questions. I think there might be a little typo and then it's good to ship!

File: app/lib/world/world.coffee (hunk 1 above)
Comment (Bryukh) [reply app/lib/world/world.coffee]: oh, hurry is a bad way. Thanks. Fixed

**Pull request merged**